README.md
# reorderable_carousel

[![pub package](https://img.shields.io/pub/v/reorderable_carousel.svg)](https://pub.dev/packages/reorderable_carousel)
[![flutter_tests](https://github.com/TNorbury/reorderable_carousel/workflows/flutter%20tests/badge.svg)](https://github.com/TNorbury/reorderable_carousel/actions?query=workflow%3A%22flutter+tests%22)
[![codecov](https://codecov.io/gh/TNorbury/reorderable_carousel/branch/main/graph/badge.svg)](https://codecov.io/gh/TNorbury/reorderable_carousel)
[![style: flutter lints](https://img.shields.io/badge/style-flutter__lints-blue)](https://pub.dev/packages/effective_dart)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

A carousel which allows you to reorder the contents within the carousel, as well as add or remove items from it.

![](https://raw.githubusercontent.com/TNorbury/reorderable_carousel/main/readme_assets/reorderable_carousel.gif)

## Getting Started

### Installing

In your Flutter project, add the package to your dependencies

```yml
dependencies:
    reorderable_carousel: ^0.4.0+1
```

### Usage example

Now you can put the carousel anywhere you want.
One thing to note: You need to have at least one item at all times

```dart
import 'package:reorderable_carousel/reorderable_carousel.dart';

//...

List<Color> colors = [Colors.blue];

//...

ReorderableCarousel(
  numItems: colors.length,
  addItemAt: (index) {
    setState(() {
      colors.insert(
          index,
          Color((Random().nextDouble() * 0xFFFFFF).toInt())
              .withOpacity(1.0));
    });
  },
  itemBuilder: (boxSize, index, isSelected) {
    return AnimatedContainer(
      key: ValueKey(colors[index]),
      duration: const Duration(milliseconds: 250),
      height: 150,
      width: boxSize,
      decoration: BoxDecoration(
        border: isSelected
          ? Border.all(color: Colors.amber, width: 10.0)
          : null,
        color: colors[index],
      ),
    );
  },
  onReorder: (oldIndex, newIndex) {
    // items have be reordered, update our list
    setState(() {
      Color swap = colors.removeAt(oldIndex);
      colors.insert(newIndex, swap);
    });
  },
  onItemSelected: (int selectedIndex) {
    // a new item has been selected
    setState(() {
      selectedColor = colors[selectedIndex];
    });
  },
),
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
### Fixed
- update version in readme

## 0.4.0 - 2022-08-23
### Changed
- Flutter 3 support and update analysis package (Thanks @jocubeit !)

## 0.3.2+1 - 2021-09-23
### Changed
- upgrade to Fluter 2.5.1

## 0.3.2 - 2021-07-12
### Added
- Allow scroll-to animation curve and duration to be specified

## 0.3.1 - 2021-04-21
### Added
- Added `draggedItemBuilder` allowing you to build the widget that's being dragged

## 0.3.0 - 2021-04-08
### Added
- #### BREAKING

    `maxNumberItems` can be used to limit the number of items in a carousel (this used to be implicitly set to 10). If not set you can add unlimited number of items.


### Changed
- addItemAt can now return a bool. Returning true or null indicates that an item was added and that the selected item should be updated. Returning false means that the item ended up not being added, and thus selected item shouldn't be updated.

    Seeing as null is an appropriate return value, this won't be a breaking change.


## 0.2.0 - 2021-03-31
### Changed
- addItemAt can now be used async. Meaning that onItemSelected won't be called until it completes

## 0.2.0-nullsafety - 2021-03-24
### Changed
- Null safety

## 0.1.1 - 2021-03-24
### Added
- Add parameter `itemWidthFraction` to control how item width is calculated

### Changed
- Change scroll calculation to accommodate non-fixed item fraction

### Fixed
- Scroll to item getting used on wrong gesture

## 0.1.0+2 - 2021-03-23
### Fixed
- Fix clicking on item not scrolling to said item

## 0.1.0+1 - 2021-03-23
### Changed
- Fix image url in readme
- Update pubspec description

## 0.1.0 - 2021-03-23
### Added
- Initial release of package

test/reorderable_carousel_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:reorderable_carousel/reorderable_carousel.dart';

void main() {
  testWidgets(
    'Tap on item',
    (WidgetTester tester) async {
      int? selectedIndex;
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              numItems: 1,
              addItemAt: (int index) {
                return null;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onItemSelected: (int idx) {
                selectedIndex = idx;
              },
              onReorder: (int oldIndex, int newIndex) {},
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      await tester.tap(find.byKey(const Key('Item 0')));
      await tester.pumpAndSettle(const Duration(seconds: 20));
      expect(selectedIndex, 0);
    },
  );

  testWidgets(
    'Drag test',
    (WidgetTester tester) async {
      int? oldIndex;
      int? newIndex;
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              numItems: 2,
              addItemAt: (int index) {
                return null;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onReorder: (int oldIdx, int newIdx) {
                oldIndex = oldIdx;
                newIndex = newIdx;
              },
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      Offset dragStart = tester.getCenter(find.byKey(const Key('Item 0')));
      Offset dragEnd = Offset(
          tester.getBottomRight(find.byKey(const Key('Item 1'))).dx + 50,
          tester.getBottomRight(find.byKey(const Key('Item 1'))).dy);

      final gesture = await tester.startGesture(
        dragStart,
        pointer: 7,
      );
      await tester.pump(const Duration(seconds: 20));
      await gesture.moveTo(dragEnd);
      await tester.pump();
      await gesture.up();
      await tester.pump();

      expect(oldIndex, 0);
      expect(newIndex, 1);
    },
  );

  testWidgets(
    'Add item callback',
    (WidgetTester tester) async {
      int? newIndex;
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              numItems: 1,
              addItemAt: (int index) {
                newIndex = index;
                return null;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onReorder: (int oldIdx, int newIdx) {},
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      await tester.tap(find.byType(IconButton).first);
      await tester.pumpAndSettle();

      expect(newIndex, 1);
    },
  );

  testWidgets(
    'Add item awaits',
    (WidgetTester tester) async {
      bool? addItemCalled;
      bool? itemSelectedCalled;
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              numItems: 1,
              addItemAt: (int index) async {
                await Future.delayed(const Duration(seconds: 1));
                addItemCalled = true;
                return null;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onReorder: (int oldIdx, int newIdx) {},
              onItemSelected: (_) {
                expect(addItemCalled, true);
                itemSelectedCalled = true;
              },
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      await tester.tap(find.byType(IconButton).first);
      await tester.pumpAndSettle(const Duration(seconds: 20));

      expect(itemSelectedCalled, true);
    },
  );

  testWidgets(
    'selected item isn\'t updated if add item returns null',
    (WidgetTester tester) async {
      bool? itemSelectedCalled = false;
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              numItems: 1,
              addItemAt: (int index) async {
                await Future.delayed(const Duration(seconds: 1));
                return false;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onReorder: (int oldIdx, int newIdx) {},
              onItemSelected: (_) {
                itemSelectedCalled = true;
              },
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      await tester.tap(find.byType(IconButton).first);
      await tester.pumpAndSettle(const Duration(seconds: 20));

      expect(itemSelectedCalled, false);
    },
  );

  testWidgets(
    'Add icon isn\'t visible if max number of items is reach',
    (WidgetTester tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              maxNumberItems: 5,
              numItems: 5,
              addItemAt: (int index) {
                return null;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onReorder: (int oldIdx, int newIdx) {},
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      expect(find.byType(IconButton), findsNothing);
    },
  );

  testWidgets(
    'Different builder used when dragging',
    (WidgetTester tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: ReorderableCarousel(
              numItems: 2,
              addItemAt: (int index) {
                return null;
              },
              itemBuilder: (double itemWidth, int index, bool isSelected) {
                return SizedBox(
                  key: Key('Item $index'),
                  height: 100,
                  width: itemWidth,
                );
              },
              onReorder: (int oldIdx, int newIdx) {},
              draggedItemBuilder: (itemWidth, index) {
                return SizedBox(
                  width: itemWidth,
                  height: 50,
                  child: Text('Dragging $index'),
                );
              },
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();

      Offset dragStart = tester.getCenter(find.byKey(const Key('Item 0')));

      final gesture = await tester.startGesture(
        dragStart,
        pointer: 7,
      );
      await tester.pump(const Duration(seconds: 20));

      // find the dragged widget
      expect(find.text('Dragging 0'), findsOneWidget);
      await gesture.up();
      await tester.pump();

      // widget should be gone after pointer up
      expect(find.text('Dragging 0'), findsNothing);
    },
  );
}


lib/src/reorderable_carousel.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

/// A carousel that can be re-ordered by dragging the elements around.
class ReorderableCarousel extends StatefulWidget {
  /// The number of items in the carousel.
  final int numItems;

  /// Callback for when the user presses the "+" button. The given index
  /// indicates where the new item should be inserted at.
  ///
  /// Returning either `true` or null indicates that a new item was added to
  /// your collection of items, and that the added item should be selected,
  /// which in turn calls [onItemSelected] (if defined) and scrolls to the that
  /// item.
  ///
  /// Returning `false` means that you didn't end up adding the item, and thus
  /// the selected item shouldn't be updated.
  final FutureOr<bool?> Function(int index) addItemAt;

  /// Builder for creating the items of the carousel.
  /// [itemWidth] indicates the maximum amount of width alloted for the item
  ///
  /// [index] is the index of the item to be built
  ///
  /// [isSelected] whether or not the item is selected. This will be true if the
  ///   item has been tapped on, or if it's currently being dragged
  ///
  /// Will be called to builded the dragged item if [draggedItemBuilder] isn't
  /// defined
  final Widget Function(double itemWidth, int index, bool isSelected)
      itemBuilder;

  /// Builder that's called when the item at [index] is being dragged.
  final Widget Function(double itemWidth, int index)? draggedItemBuilder;

  /// Called after the items have been reordered. Both [oldIndex] and [newIndex]
  /// will be > 0 and < numItems
  final void Function(int oldIndex, int newIndex) onReorder;

  /// Called whenever a new item is selected.
  final void Function(int selectedIndex)? onItemSelected;

  /// The fraction of the available width of the screen that the item will take
  /// up.
  /// The item's width will be calculated like so (a [LayoutBuilder] is used):
  /// `constraints.maxWidth / [itemWidthFraction]`
  ///
  /// Must be >= 1.0
  final double itemWidthFraction;

  /// The maximum number of items allowed in the carousel. If not set the `+`
  /// icons will never disappear.
  final int? maxNumberItems;

  /// The duration for scrolling to the next selected item.
  final Duration scrollToDuration;

  /// Animation Curve used for scrolling to the next selected item.
  final Curve scrollToCurve;

  /// Creates a new [ReorderableCarousel]
  const ReorderableCarousel({
    required this.numItems,
    required this.addItemAt,
    required this.itemBuilder,
    required this.onReorder,
    this.onItemSelected,
    this.itemWidthFraction = 3,
    this.maxNumberItems,
    this.draggedItemBuilder,
    this.scrollToDuration = const Duration(milliseconds: 350),
    this.scrollToCurve = Curves.linear,
    Key? key,
  })  : assert(numItems >= 1, "You need at least one item"),
        assert(itemWidthFraction >= 1),
        super(key: key);

  @override
  _ReorderableCarouselState createState() => _ReorderableCarouselState();
}

class _ReorderableCarouselState extends State<ReorderableCarousel> {
  bool _dragInProgress = false;

  double _itemMaxWidth = -1;
  double _startingOffset = 0;
  double _endingOffset = 0;

  // includes padding around icon button
  final double _iconSize = 24 + 16.0;
  late ScrollController _controller;
  int _selectedIdx = 0;

  final double _padding = 8.0;

  @override
  void initState() {
    super.initState();
    _controller = ScrollController();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        double width = constraints.maxWidth / widget.itemWidthFraction;
        if (width != _itemMaxWidth) {
          _itemMaxWidth = width;
          _startingOffset = (constraints.maxWidth / 2) - (_itemMaxWidth / 2);
          _endingOffset = max(0, _startingOffset - _iconSize);

          // whenever the size of this widget changes, we'll rescroll to center
          // the selected item.
          _scrollToBox(_selectedIdx);
        }

        var children = [
          SizedBox(
            width: _startingOffset,
          ),
          for (int i = 0; i < widget.numItems; i++)
            GestureDetector(
              onTap: () {
                _scrollToBox(i);
              },
              child: ConstrainedBox(
                constraints: BoxConstraints.tightFor(width: _itemMaxWidth),
                child: widget.itemBuilder(_itemMaxWidth, i, i == _selectedIdx),
              ),
            ),
          SizedBox(
            width: _endingOffset,
          ),
        ];

        return ReorderableList(
          // We want all the pages to be cached. This also
          // alleviates a problem where scrolling would get broken if
          // a page changed a position by more than ~4.
          cacheExtent: (_itemMaxWidth + _iconSize) * widget.numItems,
          controller: _controller,
          scrollDirection: Axis.horizontal,
          onReorder: (oldIndex, newIndex) {
            // compensate for the leading space
            oldIndex--;
            newIndex--;
            if (newIndex > oldIndex) {
              newIndex--;
            }

            // clamp, in the event that the reorder involves the
            // leading spaces. Removing 1 to accommodate the fact that the item
            // will be removed as part of reordering.
            newIndex = newIndex.clamp(0, widget.numItems - 1);
            widget.onReorder(oldIndex, newIndex);

            // Color swap = colors.removeAt(oldIndex);
            // colors.insert(newIndex, swap);
            setState(() {
              _dragInProgress = false;

              _updateSelectedIndex(newIndex);

              _scrollToBox(newIndex);
            });
          },
          itemCount: children.length,
          itemBuilder: (context, i) {
            return Row(
              key: ValueKey(i),
              children: [
                Listener(
                  behavior: HitTestBehavior.opaque,
                  onPointerDown: (event) {
                    _updateSelectedIndex(i - 1);

                    final list = SliverReorderableList.maybeOf(context);

                    list?.startItemDragReorder(
                      index: i,
                      event: event,
                      recognizer:
                          DelayedMultiDragGestureRecognizer(debugOwner: this),
                    );
                    _PointerSmuggler(debugOwner: this)
                      ..onStart = ((d) {
                        // the wait time has passed and the drag has
                        // started
                        setState(() {
                          _dragInProgress = true;
                        });
                        return;
                      })
                      ..addPointer(event);
                  },
                  onPointerUp: (e) {
                    setState(() {
                      _dragInProgress = false;
                    });
                  },
                  child: children[i],
                ),

                // no plus icons for the invisible boxes
                if (i != 0 && i != children.length - 1) _buildAddItemIcon(i),
              ],
            );
          },
          proxyDecorator: (child, index, animation) {
            // move and scale the dragged item
            var align = AlignmentTween(
              begin: Alignment.centerLeft,
              end: Alignment.center,
            ).animate(animation);
            var size = Tween(
              begin: 1.0,
              end: 1.1,
            ).animate(animation);

            Widget item;

            // If there is a builder for dragged items, use it. Otherwise just
            // use the regular item builder.
            if (widget.draggedItemBuilder != null) {
              item = widget.draggedItemBuilder!(
                _itemMaxWidth,
                index - 1,
              );
            } else {
              item = widget.itemBuilder(
                _itemMaxWidth,
                index - 1,
                true,
              );
            }

            return AlignTransition(
              alignment: align,
              child: ScaleTransition(
                scale: size,
                child: Material(
                  elevation: 4,
                  child: item,
                ),
              ),
            );
          },
        );
      },
    );
  }

  Widget _buildAddItemIcon(int index) {
    // once we have maxNumberItems items, don't allow anymore items to be built
    if ((widget.maxNumberItems != null &&
            widget.numItems < widget.maxNumberItems!) ||
        widget.maxNumberItems == null) {
      return AnimatedOpacity(
        opacity: _dragInProgress ? 0.0 : 1.0,
        duration: const Duration(milliseconds: 250),
        child: IconButton(
          visualDensity: VisualDensity.compact,
          icon: const Icon(Icons.add),
          onPressed: () async {
            bool? itemAdded = await widget.addItemAt(index);

            // if an item was added, or the callback didn't specify, then update
            // the selected index.
            if (itemAdded == null || itemAdded) {
              setState(() {
                _updateSelectedIndex(index);

                _scrollToBox(index);
              });
            }
          },
        ),
      );
    } else {
      return SizedBox(
        width: _iconSize,
      );
    }
  }

  void _updateSelectedIndex(int index) {
    widget.onItemSelected?.call(index);
    setState(() {
      _selectedIdx = index;
    });
  }

  void _scrollToBox(int index) {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _controller.animateTo(((_itemMaxWidth + _iconSize) * index),
          duration: widget.scrollToDuration, curve: widget.scrollToCurve);
    });
  }
}

/// This allows one pointer event to be used for multiple gesture recognizers.
/// This allows us to create a custom callback that is informed with the
/// recognizer starts, but doesn't event up actually doing anything with the
/// event (seeing as it's handled by the other gesture recognizer)
class _PointerSmuggler extends DelayedMultiDragGestureRecognizer {
  _PointerSmuggler({
    Duration delay = kLongPressTimeout,
    Object? debugOwner,
  }) : super(debugOwner: debugOwner, delay: delay);

  @override
  void rejectGesture(int pointer) {
    acceptGesture(pointer);
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:reorderable_carousel/reorderable_carousel.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  State createState() {
    return MyHomePageState();
  }
}

class MyHomePageState extends State<MyHomePage> {
  late List<Color> colors;

  late Color selectedColor;
  late int selectedIdx;

  final double iconSize = 24 + 16.0;

  @override
  void initState() {
    super.initState();
    colors = [
      Color((Random().nextDouble() * 0xFFFFFF).toInt()).withOpacity(1.0)
    ];
    selectedIdx = 0;
    selectedColor = colors[selectedIdx];
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Device Layout -- "),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        children: [
          AnimatedContainer(
            duration: const Duration(milliseconds: 250),
            width: MediaQuery.of(context).size.width / 3,
            height: MediaQuery.of(context).size.height * .6,
            color: selectedColor,
          ),
          SizedBox(
            height: MediaQuery.of(context).size.height * .25,
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 8.0),
              child: ReorderableCarousel(
                maxNumberItems: 10,
                numItems: colors.length,
                addItemAt: (index) {
                  setState(() {
                    colors.insert(
                        index,
                        Color((Random().nextDouble() * 0xFFFFFF).toInt())
                            .withOpacity(1.0));
                  });
                },
                itemBuilder: (boxSize, index, isSelected) {
                  return Column(
                    children: [
                      Expanded(
                        child: Padding(
                          padding: const EdgeInsets.all(8.0),
                          child: AnimatedContainer(
                            key: ValueKey(colors[index]),
                            duration: const Duration(milliseconds: 250),
                            height: 150,
                            width: boxSize,
                            decoration: BoxDecoration(
                              border: isSelected
                                  ? Border.all(color: Colors.amber, width: 10.0)
                                  : null,
                              color: colors[index],
                            ),
                          ),
                        ),
                      ),
                      Icon(Icons.ac_unit),
                    ],
                  );
                },
                onReorder: (oldIndex, newIndex) {
                  setState(() {
                    Color swap = colors.removeAt(oldIndex);
                    colors.insert(newIndex, swap);
                  });
                },
                onItemSelected: (int selectedIndex) {
                  setState(() {
                    selectedColor = colors[selectedIndex];
                  });
                },
                draggedItemBuilder: (itemWidth, index) {
                  return AnimatedContainer(
                    key: ValueKey(colors[index]),
                    duration: const Duration(milliseconds: 250),
                    height: 150,
                    width: itemWidth,
                    decoration: BoxDecoration(
                      border: Border.all(color: Colors.amber, width: 10.0),
                      color: colors[index],
                    ),
                  );
                },
              ),
            ),
          )
        ],
      ),
    );
  }
}
